File: go/backtracking/39_combination_sum.go
package backtracking

// combinationSum 组合总和
func combinationSum(candidates []int, target int) [][]int {

	res := make([][]int, 0)
	path := make([]int, 0)

	// backtrack 回溯，start_idx 为了避免重复剪枝 比如[3,4,5],target=9 中[4,5]与[5,4]是一样的
	var backtrack func(startIdx int, total int)
	backtrack = func(startIdx, total int) {

		if total > target {
			return
		}

		if total == target {
			p := make([]int, len(path))
			copy(p, path)
			res = append(res, p)
			return
		}

		// 剪枝不选择之前选择过的元素了
		for i := startIdx; i < len(candidates); i++ {
			path = append(path, candidates[i])
			// 递归， 因为同一个元素可以重复选择，所以startIdx 为i
			backtrack(i, total+candidates[i])
			// 回溯
			path = path[:len(path)-1]
		}

	}

	backtrack(0, 0)
	return res
}

File: rust/src/backtracking/_39_combination_sum.rs
/// 组合总和
pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    let mut res = vec![];
    let mut path = vec![];

    let mut candidates = candidates;

    fn backtrack(
        start_idx: usize,
        mut total: i32,
        target: i32,
        path: &mut Vec<i32>,
        res: &mut Vec<Vec<i32>>,
        candidates: &[i32],
    ) {
        if total > target {
            return;
        }

        if total == target {
            res.push(path.clone());
            return;
        }

        for i in start_idx..candidates.len() {
            // 如果candidates 是排好序的，剪枝
            if total > target {
                break;
            }

            path.push(candidates[i]);
            backtrack(i, total + candidates[i], target, path, res, candidates);
            // 回溯
            path.pop();
        }
    }

    candidates.sort();
    backtrack(0, 0, target, &mut path, &mut res, &candidates);
    res
}
